api/src/auth_routes.ts

```diff
@@ -77,37 +104,87 @@ export function add_auth_routes(app: any, handlers: any) {
       providers: available_provider_info,
       localAuth: true, // maybe make this configurable?
       messages: req.flash(),
+      redirect: redirect,
     });
   });
 
-  // handle local login post request
-  app.post(
-    '/auth/local',
-    passport.authenticate('local', {
-      successRedirect: '/send-token',
-      failureRedirect: '/auth',
-    })
-  );
+  app.get('/logout/', (req: any, res: any, next: any) => {
+    const redirect = validateRedirect(req.query?.redirect || '/');
+
+    if (req.user) {
+      req.logout((err: any) => {
+        if (err) {
+          return next(err);
+        }
+      });
+    }
+    res.redirect(redirect);
+  });
+
+  const authenticate_return = (
+    req: any,
+    res: any,
+    next: any,
+    redirect: string
+  ) => {
+    return (err: any, user: any) => {
+      if (err) {
+        return next(err);
+      }
+      if (!user) {
+        req.flash('message', 'Invalid username or password');
+        return res.redirect('/auth/?redirect=' + redirect);
+      }
+
+      req.login(user, async (loginErr: any) => {
+        if (loginErr) {
+          return next(loginErr);
+        }
+        return redirect_with_token(res, user, redirect);
+      });
+    };
+  };
+
+  const redirect_with_token = async (
+    res: any,
+    user: Express.User,
+    redirect: string
+  ) => {
+    // Generate a token
+    const token = await generateUserToken(user);
+    // Append the token to the redirect URL
+    const redirectUrlWithToken = `${redirect}?token=${token.token}`;
+
+    // Redirect to the app with the token
+    return res.redirect(redirectUrlWithToken);
+  };
+
+  app.post('/auth/local', (req: any, res: any, next: any) => {
+    const redirect = validateRedirect(req.query?.redirect || '/');
+    passport.authenticate(
+      'local',
```

Comment:
doc strings/types etc for this function - and what is the significance of local here?